ports/esp32/machine_pwm.c

```diff
@@ -87,32 +141,117 @@ STATIC int set_freq(int newval) {
-// MicroPython bindings for PWM
+#define SAME_FREQ_ONLY (true)
+#define SAME_FREQ_OR_FREE (false)
+#define ANY_MODE (-1)
+// Return timer_idx. Use TIMER_IDX_TO_MODE(timer_idx) and TIMER_IDX_TO_TIMER(timer_idx) to get mode and timer
+STATIC int find_timer(int freq, bool same_freq_only, int mode/*=ANY_MODE*/) {
+    int free_timer_idx_found = -1;
+    // Find a free PWM Timer using the same freq
+    for (int timer_idx = 0; timer_idx < PWM_TIMER_MAX; ++timer_idx) {
+        if ((mode == ANY_MODE) || (mode == TIMER_IDX_TO_MODE(timer_idx))) {
+            if (timers[timer_idx].freq_hz == freq) {
+                // A timer already uses the same freq. Use it now.
+                return timer_idx;
+            }
+            if (!same_freq_only && (free_timer_idx_found == -1) && (timers[timer_idx].freq_hz == -1)) {
+                free_timer_idx_found = timer_idx;
+                // Continue to check if a channel with the same freq is in use.
+            }
+        }
+    }
 
-typedef struct _machine_pwm_obj_t {
-    mp_obj_base_t base;
-    gpio_num_t pin;
-    uint8_t active;
-    uint8_t channel;
-} machine_pwm_obj_t;
+    return free_timer_idx_found;
+}
+
+// Return true if the timer is in use in addition to current channel
+STATIC bool is_timer_in_use(int current_channel_idx, int timer_idx) {
+    for (int i = 0; i < PWM_CHANNEL_MAX; ++i) {
+        if ((i != current_channel_idx) && (chans[i].timer_idx == timer_idx)) {
+            return true;
+        }
+    }
+
+    return false;
+}
+
+// Find a free PWM channel, also spot if our pin is already mentioned.
+// Return channel_idx. Use CHANNEL_IDX_TO_MODE(channel_idx) and CHANNEL_IDX_TO_CHANNEL(channel_idx) to get mode and channel
+STATIC int find_channel(int pin, int mode) {
+    int avail_idx = -1;
+    int channel_idx;
+    for (channel_idx = 0; channel_idx < PWM_CHANNEL_MAX; ++channel_idx) {
+        if ((mode == ANY_MODE) || (mode == CHANNEL_IDX_TO_MODE(channel_idx))) {
+            if (chans[channel_idx].pin == pin) {
+                break;
+            }
+            if ((avail_idx == -1) && (chans[channel_idx].pin == -1)) {
+                avail_idx = channel_idx;
+            }
+        }
+    }
+    if (channel_idx >= PWM_CHANNEL_MAX) {
+        channel_idx = avail_idx;
+    }
+    return channel_idx;
+}
+
+/******************************************************************************/
+// MicroPython bindings for PWM
 
 STATIC void mp_machine_pwm_print(const mp_print_t *print, mp_obj_t self_in, mp_print_kind_t kind) {
     machine_pwm_obj_t *self = MP_OBJ_TO_PTR(self_in);
-    mp_printf(print, "PWM(%u", self->pin);
+    mp_printf(print, "PWM(pin=%u", self->pin);
     if (self->active) {
-        mp_printf(print, ", freq=%u, duty=%u", timer_cfg.freq_hz,
-            ledc_get_duty(PWMODE, self->channel));
+        int duty = get_duty(self);
+        mp_printf(print, ", freq=%u, duty=%u(%.1f%%)", ledc_get_freq(self->mode, self->timer),
```

Comment:
please don't include the percentage in the duty value, just include the raw value, to match how it would be constructed in Python